src/Microsoft.ML.ImageAnalytics/ImageGrayscale.cs

```diff
@@ -218,6 +218,9 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
                             g.DrawImage(src, srcRectangle, 0, 0, src.Width, src.Height, GraphicsUnit.Pixel, attributes);
                         }
                         Contracts.Assert(dst.Width == src.Width && dst.Height == src.Height);
+
+                        src.Dispose();
```

Comment:
Is Disposing it the correct thing to do here? Are you given ownership of the `src` object? If so, then it is fine. But if `getSrc` caches that object and tries using it, people will get an ObjectDisposedException if they try to use it.

Also - there is a case on line 210 where this method could return early without disposing `src`.

Reply:
I’m aware about line 210 but that is an error condition and not sure bitmap object has the ownership of the file but I’ll check. I did also consider the caching issue and didn’t see it being cached anywhere. None of the image transforms seem to cache the bitmap object.

Reply:
ML.NET pipeline runs in a read-only mode. `Disposing something` means read-and-write.

Reply:
You need to dispose otherwise it is a bug. Not entirely sure what your argument here is but we need a solution.

Reply:
Sure. We need a solution. My question is more, why the existing solution and infrastructure is insufficient, since *in principle* the `src` should be disposed by the getter, then *finally* disposed when the row cursor is itself disposed. I am wondering what has broken down. #3327 is not sufficient.

---
In reply to: [275461003](https://github.com/dotnet/machinelearning/pull/3328#discussion_r275461003) [](ancestors = 275461003)